indico/legacy/pdfinterface/latex.py

```diff
@@ -38,6 +45,45 @@
 from indico.util.string import render_markdown
 
 
+#: A rate limiter for PDF generation endpoints that are available publicly without logging in
+latex_rate_limiter = LocalProxy(functools.cache(lambda: make_rate_limiter('latex', config.LATEX_RATE_LIMIT)))
+cache = make_scoped_cache('latex-pdfs')
+
+
+def generate_cached_pdf(fn, key, obj=None) -> BytesIO:
+    """Generate a PDF from LaTeX with caching and rate limiting.
+
+    This expects a callable (which will be called with `obj` as its argument)
+    that generates a PDF file and returns the PDF bytes.
+
+    It also takes a `key` which should be unique to this particular PDF generation and
+    of course the `obj` which is the object from which the PDF is generated (this must be
+    an SQLAlchemy object).
+
+    The generated PDF is cached (longer when the request comes from an unauthenticated user),
+    and rate limiting is applied as well if the user is unauthenticated.
+
+    In cases where there is no single object, it can be omitted, in which case `fn` will be
+    called without arguments. The `key` must contain enough details to ensure proper cache
+    separation in that case.
```

Comment:
I'd switch the order of the last two paragraphs and make it clearer what the `it` refers to: 

```suggestion
    In cases where there is no single object, the `obj` argument can be omitted, in which case `fn` will be
    called without arguments. The `key` must contain enough details to ensure proper cache
    separation in that case.

    The generated PDF is cached (longer when the request comes from an unauthenticated user),
    and rate limiting is applied as well if the user is unauthenticated.
```

Not so sure about the the function signature :thinking:  What if we only allow passing the key and not the object? This would force the caller to construct the key, which we already do in two places anyway, but it could simplify the code (and the docstring) a bit.

Reply:
I want to make it less likely to shoot yourself in the foot there (not including an identifier for the object in the cache key and thus accidentally sharing a single cache between different object)